python/essex-config/README.md

```diff
@@ -0,0 +1,195 @@
+# essex-configuration
+
+essex-configuration is a Python library for creating configuration objects that can read from various sources, including files, environment variables, and Azure Key Vault.
+
+## Installation
+
+To install the essex-configuration library, use pip:
+
+`pip install essex-configuration`
+
+## Basic Usage
+
+Here's how to create a configuration object for connecting to a customer database:
+
+
+```python
+from typing import Annotated, ClassVar
+from essex_config import Config, ConfigurationField, FieldVisibility
+
+
+class CustomerDatabase(Config):
+    """Class that holds the configuration to connect to the Customer Database"""
+
+    host: Annotated[str, ConfigurationField(
+        description="Host ip or url to connect to the database",
+    )] = "127.0.0.1"
+
+    port: Annotated[int, ConfigurationField(
+        description="Port to connect to the database",
+    )]
+    
+    password: Annotated[str, ConfigurationField(
+        field_visibility=FieldVisibility.SECRET, description="Password to connect to sql database"
+    )]
+
+
+if __name__ == "__main__":
+    config = CustomerDatabase.get_config()
+    print(config)
+```
+
+When `CustomerDatabase.get_config()` is executed, the values will be populated from environment variables or default values.
+
+
+## Advanced Usage
+
+
+### Sources
+
+You can add different sources for configuration data by defining the `sources` class variable:
+
+```python
+class MyConfig(Config):
+    sources: ClassVar[list[Source]] = ...
+```
+
+`essex-configuration` supports three built-in sources:
+
+1. `EnvSource()`: Default source, reads from environment variables.
+2. `FileSource(file_path: Path | str, use_env_var: bool = False)`: Reads from a toml, json, or yaml file. use_env_var=True allows the file path to be specified via an environment variable.
+3. `KeyvaultSource(keyvault_name: str, use_env_var: bool = False)`: Fetches values from an Azure Key Vault. use_env_var=True allows the Key Vault name to be specified via an environment variable.
+
+Example of multiple sources:
+
+```python
+from typing import Annotated, ClassVar
+from essex_config import Config, ConfigurationField, FieldVisibility
+
+class CustomerDatabase(Config):
+    """Configuration for connecting to the Customer Database"""
+
+    sources: ClassVar[list[Source]] = [
+        EnvSource(),
```

Comment:
Should we allow this to have something like this: `EnvSource(prefix="SOMETHIG_")`. So we can add a suffix bound to our application to the environment variables?